proposals/yield-in-try-catch.md

```diff
@@ -0,0 +1,500 @@
+5. The `Dispose` method will mirror `catch` and `finally` blocks in the same way it mirrors `finally` blocks today and dispatch to the appropriate method
+
+The exception object will be lifted into a field just as any other local would be and accesses to it in the generated `catch` and `when` methods will use the field. In the case the language allowed `when` clauses the pattern variables would be lifted into fields as well.
+
+For example consider this code sample:
+
+```csharp
+static IEnumerable<int> Iterator()
+{
+    try
+    {
+        try
+        {
+            yield return M(1);
+        }
+        finally
+        {
+            Console.WriteLine("Finally Inner");
+        }
+    }
+    catch (InvalidOperationException ex1) when (ex.Message.Contains("hello"))
+    {
+        Console.WriteLine("Catch1");
+    }
+    catch (Exception ex2) 
+    {
+        Console.WriteLine("Catch2");
+    }
+    finally
+    {
+        Console.WriteLine("Finally Outer");
+    }
+}
+static T M<T>(T t) => t;
+```
+
+Will generate the following `Dispose` method:
+
+```csharp
+[DebuggerHidden]
+void IDisposable.Dispose()
+{
+    int num = <>1__state;
+    if ((uint)(num - -4) > 1u && num != 1)
+    {
+        return;
+    }
+    try
+    {
+        if (num != -4 && num != 1)
+        {
+            return;
+        }
+        try
+        {
+        }
+        finally
+        {
+            <>m__Finally2();
+        }
+    }
+    catch (InvalidOperationException ex) when (<>1__ex1 = ex, <>m__When1())
+    {
+        <>m__Catch1();
+    }
+    catch (Exception ex)
+    {
+        <>1__ex2 = ex;
+        <>m__Catch1();
+    }
+    finally
+    {
+        <>m__Finally1();
+    }
+}
+```
+
+The `<>1__ex1 = ex` in the `when` clause is not legal but the IL generated for the `when` will conceptually have this behavior.
+
+This would add a bit of complexity to the feature and it only produces observable differences when a `finally` block throws an exception on the `Dispose` path. That is likely a rare case. Further it potentially increases the complexity for developers. The `Dispose` method at first glance is likely unintuitive in that it only mirrors `finally` blocks and no other statements but that is also a very simple rule to learn. Preserving `catch` and `finally` and discussing how the code flows between them is potentially more complex.
+
+This also brings into question what happens when a `yield` occurs in a `catch` during `Dispose`. That cannot execute correctly as the state machine can't suspend during `Dispose`. To account for this the feature likely needs further restrictions like:
+
+- The `yield` statement will be allowed in a `catch` block provided that:
+  - The `try` block does not contain a `finally` block.
+  - The `try` block does not have a nested `finally` block.
+
+On the whole it seems simpler to not support `catch` in `Dispose` paths.
+
+## Related Issues
+
+Related Items:
+
+- https://github.com/dotnet/csharplang/discussions/765
```

Comment:
Could be good to link to other duplicate items:
- https://github.com/dotnet/csharplang/discussions/6819
- https://github.com/dotnet/csharplang/discussions/8994
- https://github.com/dotnet/docs/issues/42112